seq2simulate/run_make_proviral_data.py

```diff
@@ -0,0 +1,483 @@
+# hypermutation_rate = 3
+
+def get_diffs1(seq0, seq1):
+    """
+    Simply return a list of 1-based different indices between two strings.
+    """
+    assert len(seq0) == len(seq1)
+    return [i for i in range(len(seq0)) if seq0[i] != seq1[i]]
+
+
+def run_proviral(sequences_path, working_dir, out_dir, platform, paired_end, unclean_working=False, hypermutation_rate=3):
+    """
+
+    Perform proviral mutations (including hypermutations, deletions,
+    insertions, frameshifts, inserting stop-codons, and inversions) on a set of
+    sequences and generate an NGS dataset for viral and proviral sequences, then
+    sample these two FASTQ files to create a new FASTQ file consisting of a
+    fraction of reads produced from each proviral (and viral) dataset.
+
+    Args:
+        sequences_path: path to FASTA sequence file
+        working_dir: temporary directory for storing output files 
+        out_dir: directory in which to store result
+        platform: platform (e.g. roche)
+        paired_end: produce paired_end data
+        unclean_working: do not delete working directory files upon completion
+        hypermutation_rate: rate of hypermutation per 100 bp
+
+    Returns:
+        True 
+
+
+    """
+    print('Using temporary working directory: {}'.format(working_dir))
+
+    sequences = [s for s in SeqIO.parse(sequences_path, 'fasta')]
+
+    data_files, diff_files = _make_mutation_data_files(sequences, working_dir, hypermutation_rate=hypermutation_rate)
+    data_files['null'] = sequences_path
+
+    fastq_files, sam_files = _make_art_files(data_files, platform, working_dir, paired_end=paired_end)
+
+    open_fq_files, open_sam_files, open_diff_files = _open_all_data_files(sequences, fastq_files, sam_files, diff_files)
+
+    n_reads = min([len(j) for i in open_fq_files.values() for j in i])
+    n_reads_frac = int(1.0/len(MUTATIONS) * n_reads)
+    n_mutated_reads = {
+        'null': n_reads_frac,
+        'hypermutation': n_reads_frac,
+        'longdel': n_reads_frac,
+        'insertion': n_reads_frac,
+        'frameshift': n_reads_frac,
+        'stopcodon': n_reads_frac,
+        'inversion': n_reads_frac,
+    }
+
+    for i,j in n_mutated_reads.iteritems():
+        print('Sampling {0} ({1}) reads out of {2} total reads (r={3:.3f}).'.format(j, i, n_reads, j/float(n_reads)))
+
+    # NOTE: 
+    # roche: fq sequence is fwd strand, sam seq is fwd strand
+    # illumina: fq sequence is rev comp of fwd strand, sam seq is fwd strand
+    #           i.e. fq has complement of 3'->5' end of fwd strand, sam has complement of this
+    # illumina+paired: 
+    #           fq1 sequence is rev comp of fwd strand, sam seq is fwd strand
+    #           i.e. fq has complement of 3'->5' end of fwd strand, sam has complement of this
+    #           fq2 sequence is fwd strand, sam seq is fwd strand
+
+    fastq_samples = {}
+    for mutation_type, fastq_files in open_fq_files.iteritems():
+        if len(fastq_files) == 1:
```

Comment:
Pretty sure you can completely collapse this switch with a variation on `for file in fastq_files`